File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        queue = deque()
        maxWidth = 1
        
        queue.append((root, 0))
        
        
        while queue:
            start = queue[0][1]
            end = queue[-1][1]
            
            levelSize = len(queue)
            # print(start, end)
            maxWidth = max(maxWidth, end - start + 1)
            
            for i in range(levelSize):
                currNode = queue.popleft()
                
                idx = currNode[1] - start
                
                if currNode[0].left != None:
                    queue.append((currNode[0].left, 2 * idx))
                
                if currNode[0].right != None:
                    queue.append((currNode[0].right, 2 * idx + 1))
        return maxWidth
            

File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        return solve(root);
    }
private:
    int solve(TreeNode* root) {
        int maxWidth = 1;
        
        queue<pair<TreeNode*, long long>> queue;
        queue.push({root, 0});
        
        while(!queue.empty()) {
            int levelSize = queue.size();
            int start = queue.front().second;
            int end = queue.back().second;
            
            maxWidth = max(maxWidth, end - start + 1);
                
            while(levelSize--) {
                auto currNode = queue.front();
                queue.pop();
                
                int idx = currNode.second - start;
                
                if(currNode.first -> left != nullptr)
                    queue.push({currNode.first -> left, (long long)2 * idx});
                if(currNode.first -> right != nullptr)
                    queue.push({currNode.first -> right, (long long)2 * idx + 1});
            }
        }
        return maxWidth;
        
    }
};